### PR TITLE
feat(kafka-logger): support for specified the log formats via admin API.

### PR DESCRIPTION
**Pull request by chzhuo**
### What this PR does / why we need it:
as title.

For example:

```lua
curl http://****/apisix/admin/plugin_metadata/kafka-logger -d '
{
    "log_format": {
        "host": "$host",
        "@timestamp": "$time_iso8601",
        "client_ip": "$remote_addr"
    }
}'
```

when we enabled plugin http-logger, we will get the message body like:

```http
{"host":"localhost","@timestamp":"2020-09-23T18:29:07-04:00","client_ip":"127.0.0.1","route_id":"1"}
{"host":"localhost","@timestamp":"2020-09-23T18:29:07-04:00","client_ip":"127.0.0.1","route_id":"1"}
```

### Pre-submission checklist:

* [x] Did you explain what problem does this PR solve? Or what new features have been added?
* [x] Have you added corresponding test cases?
* [x] Have you modified the corresponding document?
* [x] Is this PR backward compatible? **If it is not backward compatible, please discuss on the [mailing list](https://github.com/apache/apisix/tree/master#community) first**

### CODE REVIEW

File: apisix/plugins/kafka-logger.lua [hunk 1]
```diff
@@ -152,8 +166,15 @@ function _M.log(conf, ctx)
...
-        entry = log_util.get_full_log(ngx, conf)
-        core.log.info("full log entry: ", core.json.delay_encode(entry))
+        local metadata = plugin.plugin_metadata(plugin_name)
+        core.log.info("metadata: ", core.json.delay_encode(metadata))
+        if metadata and metadata.value.log_format 
+            and core.table.nkeys(metadata.value.log_format) > 0 then
+            entry = log_util.get_custom_format_log(metadata.value.log_format)
+        else
+            entry = log_util.get_full_log(ngx, conf)
+            core.log.info("full log entry: ", core.json.delay_encode(entry))
```
Comment (tokers) [inline apisix/plugins/kafka-logger.lua:176]: Should also print this log for the "custom_format_log".

File: docs/zh/latest/plugins/kafka-logger.md [hunk 2]
```diff
@@ -171,6 +171,33 @@ HTTP/1.1 200 OK
 hello, world
 ```
 
+## 插件元数据设置
+
+| 名称             | 类型    | 必选项 | 默认值        | 有效值  | 描述                                             |
+| ---------------- | ------- | ------ | ------------- | ------- | ------------------------------------------------ |
+| log_format       | object  | 可选   |               |         | 以 Hash 对象方式声明日志格式。对 value 部分，仅支持字符串。如果是以`$`开头，则表明是要获取 [Nginx 内置变量](http://nginx.org/en/docs/varindex.html)。特别的，该设置是全局生效的，意味着指定 log_format 后，将对所有绑定 kafka-logger 的 Route 或 Service 生效。 |
```
Comment (tokers) [inline docs/zh/latest/plugins/kafka-logger.md:178]: `该设置是全局生效的` can be set to bold, to highlight  this message.

File: t/plugin/kafka-logger.t [hunk 3]
```diff
@@ -722,3 +722,43 @@ GET /t
...
+            ngx.status = code
+            ngx.say(body)
+        }
+    }
+--- request
+GET /t
+--- response_body
+passed
+--- no_error_log
+[error]
```
Comment (tokers) [inline t/plugin/kafka-logger.t:764]: Please check the editor, the EOL symbol is required.

File: t/plugin/kafka-logger.t [hunk 4]
```diff
@@ -722,3 +722,43 @@ GET /t
 [qr/partition_id: 1/,
 qr/partition_id: 0/,
 qr/partition_id: 2/]
+
+
+
+=== TEST 20: add plugin metadata
```
Comment (tokers) [inline t/plugin/kafka-logger.t:728]: Should add another test case to cover whether the `log_format` setting is effective.

File: apisix/utils/log-util.lua [hunk 5]
```diff
@@ -17,10 +17,58 @@
...
+            log_format[k] = {true, var_name:sub(2)}
+        else
+            log_format[k] = {false, var_name}
+        end
+    end
+    core.log.info("log_format: ", core.json.delay_encode(log_format))
+    return log_format
+end
+
+local function get_custom_format_log(format)
```
Comment (spacewander) [inline apisix/utils/log-util.lua:51]: Better to pass ctx as an argument

File: apisix/plugins/http-logger.lua [hunk 6]
```diff
@@ -198,23 +168,10 @@ function _M.log(conf, ctx)
...
-            end
-        end
 
-        local matched_route = ctx.matched_route and ctx.matched_route.value
-        if matched_route then
-            entry.service_id = matched_route.service_id
-            entry.route_id = matched_route.id
-        end
+    if metadata and metadata.value.log_format
+        and core.table.nkeys(metadata.value.log_format) > 0 then
```
Comment (spacewander) [inline apisix/plugins/http-logger.lua:173]:
```suggestion
      and core.table.nkeys(metadata.value.log_format) > 0
    then
```

File: apisix/plugins/kafka-logger.lua [hunk 7]
```diff
@@ -152,8 +166,16 @@ function _M.log(conf, ctx)
         -- core.log.info("origin entry: ", entry)
 
     else
-        entry = log_util.get_full_log(ngx, conf)
-        core.log.info("full log entry: ", core.json.delay_encode(entry))
+        local metadata = plugin.plugin_metadata(plugin_name)
+        core.log.info("metadata: ", core.json.delay_encode(metadata))
+        if metadata and metadata.value.log_format
+            and core.table.nkeys(metadata.value.log_format) > 0 then
```
Comment (spacewander) [inline apisix/plugins/kafka-logger.lua:172]:
```suggestion
          and core.table.nkeys(metadata.value.log_format) > 0
        then
```

File: docs/zh/latest/plugins/http-logger.md [hunk 8]
```diff
@@ -89,7 +89,7 @@ hello, world
 
 | 名称             | 类型    | 必选项 | 默认值        | 有效值  | 描述                                             |
 | ---------------- | ------- | ------ | ------------- | ------- | ------------------------------------------------ |
-| log_format       | object  | 可选   | {"host": "$host", "@timestamp": "$time_iso8601", "client_ip": "$remote_addr"} |         | 以 JSON 对象方式声明日志格式。对 value 部分，仅支持字符串。如果是以 `$` 开头，则表明是要获取 __APISIX__ 变量或 [Nginx 内置变量](http://nginx.org/en/docs/varindex.html)。特别的，该设置是全局生效的，意味着指定 log_format 后，将对所有绑定 http-logger 的 Route 或 Service 生效。 |
+| log_format       | object  | 可选   | {"host": "$host", "@timestamp": "$time_iso8601", "client_ip": "$remote_addr"} |         | 以 JSON 对象方式声明日志格式。对 value 部分，仅支持字符串。如果是以 `$` 开头，则表明是要获取 __APISIX__ 变量或 [Nginx 内置变量](http://nginx.org/en/docs/varindex.html)。特别的，**该设置是全局生效的**，意味着指定 log_format 后，将对所有绑定 http-logger 的 Route 或 Service 生效。 |
```
Comment (spacewander) [inline docs/zh/latest/plugins/http-logger.md:92]: Could you also update the English version of doc? You can use an online translator if you need help. Thanks!

**Review by spacewander — APPROVED** [review]

**Review by moonming — APPROVED** [review]

**Pull request merged**